aten/src/ATen/native/Indexing.cpp

```diff
@@ -0,0 +1,220 @@
+      for (int64_t j = 0; j < nonzero.size(1); j++) {
+        result.emplace_back(nonzero.select(1, j));
+      }
+    } else {
+      result.emplace_back(index);
+    }
+  }
+  return result;
+}
+
+static bool hasContiguousSubspace(TensorList tl) {
+  // true if all the non-null tensors are adjacent
+  auto isDefined = [](const Tensor & tensor){ return tensor.defined(); };
+  auto isNull = [](const Tensor & tensor){ return !tensor.defined(); };
+  auto start = std::find_if(tl.begin(), tl.end(), isDefined);
+  auto stop = std::find_if(tl.rbegin(), tl.rend(), isDefined);
+  auto it = std::find_if(start, stop.base(), isNull);
+  return it == stop.base();
+}
+
+// Transposes the tensor and indices together so that all the non-null indices
+// index the first k dimensions of the tensor. Returns the transposed tensor
+// and the reordered indices. For example:
+//  transposeToFront(tensor, {nullptr, a, nullptr, b})
+// returns
+//  tensor.permute([1, 3, 0, 2]), {a, b, nullptr, nullptr}
+static std::tuple<Tensor, std::vector<Tensor>>
+transposeToFront(Tensor self, TensorList indices) {
+  std::vector<int64_t> dims;
+  std::vector<Tensor> transposedIndices;
+  dims.reserve(self.dim());
+  for (int64_t i = 0; i < self.dim(); i++) {
+    if (indices[i].defined()) {
+      dims.push_back(i);
+      transposedIndices.emplace_back(indices[i]);
+    }
+  }
+  for (int64_t i = 0; i < self.dim(); i++) {
+    if (!indices[i].defined()) {
+      dims.push_back(i);
+      transposedIndices.emplace_back();
+    }
+  }
+  return std::make_tuple<>(self.permute(dims), std::move(transposedIndices));
+}
+
+static std::vector<int64_t> computeLinearStride(const Tensor & tensor) {
+  // computes the stride as if tensor were contigous
+  auto sizes = tensor.sizes();
+  std::vector<int64_t> stride(tensor.dim());
+  stride[tensor.dim() - 1] = 1;
+  std::partial_sum(sizes.rbegin(), sizes.rend() - 1, stride.rbegin() + 1, std::multiplies<int64_t>());
+  return stride;
+}
+
+// Unsqueezes src `before` times at the front and `after` times at the end
+static Tensor unsqueezeN(const Tensor & src, int64_t before, int64_t after) {
+  auto srcSizes = src.sizes();
+  auto nDim = src.dim();
+  std::vector<int64_t> sizes(nDim + before + after, 1);
+  for (int64_t i = 0; i < nDim; i++) {
+    sizes[i + before] = srcSizes[i];
+  }
+  return src.view(sizes);
+}
+
+static Tensor computeLinearIndex(const Tensor & src, TensorList indices) {
+  auto strides = computeLinearStride(src);
+
+  // Compute the linear index by multiplying the indexing tensors by the
+  // stride and summing them. All the indexing tensors have the same shape at
+  // this point. We also compute the number of dimensions before and after that
+  // are not being index.
+  Tensor linearIndex;
+  int64_t emptyBefore = 0, emptyAfter = 0, nElemBefore = 1, nElemAfter = 1;
+  for (int64_t i = 0; i < src.dim(); i++) {
+    if (indices[i].defined()) {
+      Tensor index = indices[i] * strides[i];
```

Comment:
How does this operator overload work? From my interpretation, if `indices` is on the GPU this will invoke a `cudaMemcpy` and `mm` kernel launch for each dimension correct?

Reply:
It's multiplying the indices (a LongTensor, maybe CUDA) by an integer. It translates to `mul` at the TH-level. So just a single kernel launch and no extra memcpy.

Reply:
I believe `strides[i]` is a number, so it performs a `cmul`, but indeed, there is a kernel launch for every dimension.